src/rpc/blockchain.cpp

```diff
@@ -1645,34 +1646,141 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
     ret.pushKV("minrelaytxfee", ValueFromAmount(::minRelayTxFee.GetFeePerK()));
     ret.pushKV("unbroadcastcount", uint64_t{pool.GetUnbroadcastTxs().size()});
+
+    if (feeLimits) {
+        const MempoolHistogramFeeRates& limits{feeLimits.value()};
+
+        /* Keep histogram per...
+         * ... cumulated tx sizes
+         * ... txns (count)
+         * ... cumulated fees */
+        std::vector<uint64_t> sizes(limits.size(), 0);
+        std::vector<uint64_t> count(limits.size(), 0);
+        std::vector<CAmount> fees(limits.size(), 0);
+
+        for (const CTxMemPoolEntry& e : pool.mapTx) {
+            const CAmount fee{e.GetFee()}, afees{e.GetModFeesWithAncestors()}, dfees{e.GetModFeesWithDescendants()};
+            const uint32_t size{(uint32_t)e.GetTxSize()}, asize{(uint32_t)e.GetSizeWithAncestors()}, dsize{(uint32_t)e.GetSizeWithDescendants()};
+
+            const CAmount fpb{CFeeRate{fee, size}.GetFee(1)};     // Fee rate per byte
+            const CAmount afpb{CFeeRate{afees, asize}.GetFee(1)}; // Fee rate per byte including ancestors
+            const CAmount dfpb{CFeeRate{dfees, dsize}.GetFee(1)}; // Fee rate per byte including descendants
+
+            // Fee rate per byte including ancestors & descendants
+            const CAmount tfpb{CFeeRate{afees + dfees - fee, asize + dsize - size}.GetFee(1)};
+            const CAmount fee_rate_per_byte{std::max(std::min(dfpb, tfpb), std::min(fpb, afpb))};
```

Comment:
Can you explain these two lines?

```C++
1670    const CAmount tfpb{CFeeRate{afees + dfees - fee, asize + dsize - size}.GetFee(1)};
1671    const CAmount fee_rate_per_byte{std::max(std::min(dfpb, tfpb), std::min(fpb, afpb))};
```

Reply:
I'm confused by these as well.

This breaks down when there are multiple descendants or ancestors in various scenarios (all txs assumed to be same size). 

If you e.g. had a parent transaction paying 5 sat/vB that had a child paying 1 sat/vB, the child would be irrelevant for the parent's effective feerate and vice versa. If you however have a parent paying 3 sat/vB, and two children where one pays 5 sat/vB and one pays 1 sat/vB. The one with the 5 sat/vB would form a CPFP situation with the parent, while the other child would be effectively unrelated: since the parent's feerate is larger than the child's, the ancestor will be included before the child independently. Only children pay for parents, parents don't pay for children in this case. ;)

Beyond that, many transactions will have overlapping ancestries or descendants, which means that the bucketing across the complete graph of related transactions would often count related txs multiple times, e.g. where a parent paying 3 sat/vB has two children paying 4 sat/vB and 7 sat/vB, only the latter will form a CPFP and the parent will already be included when the child with 4 sat/vB is up for transaction selection.

Reply:
I can only say this: https://github.com/bitcoin/bitcoin/pull/21422#discussion_r754603944 and that I will try to fix it.